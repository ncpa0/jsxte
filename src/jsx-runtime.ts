export * from "./jsx/jsx.types";
export * from "./jsx/jsx-runtime";
