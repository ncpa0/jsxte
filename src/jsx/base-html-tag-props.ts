import type { Rewrap } from "../html-parser/types";

export type AttributeBool = true | false | "true" | "false";

export type HTMLProps<T extends object = never> = Rewrap<
  ExtendBaseProps<
    [T] extends [never]
      ? JSXTE.BaseHTMLTagProps
      : Partial<T> & JSXTE.BaseHTMLTagProps
  >
>;

type ExtendBaseProps<P> = {
  [K in keyof P]: JSXTE.AttributeAcceptedTypes extends {
    [E in K]: infer T;
  }
    ? T | P[K]
    : P[K];
};

declare global {
  namespace JSXTE {
    interface AttributeAcceptedTypes {}

    export type TagElement = {
      type: "tag";
      tag:
        | string
        | ((props: ElementProps) => Element)
        | ((props: ElementProps) => Promise<Element>);
      props: ElementProps;
    };

    export type TextNodeElement = {
      type: "textNode";
      text: string;
    };

    export type SyncElement = TagElement | TextNodeElement;

    type ElementChildren =
      | JSX.Element
      | string
      | number
      | Array<
          JSX.Element | string | number | Array<JSX.Element | string | number>
        >;

    type ElementProps = {
      children?: ElementChildren;
      [k: string]: any;
    };

    type PropsWithChildren<P extends object> = P & {
      children?: JSXTE.ElementChildren;
    };

    type Component<P extends object = {}> = (
      props: PropsWithChildren<P>
    ) => JSX.Element;

    type AsyncComponent<P extends object = {}> = (
      props: PropsWithChildren<P>
    ) => Promise<JSX.Element>;

    export interface BaseHTMLTagProps {
      children?: ElementChildren;

      accesskey?: string;
      class?: string;
      contenteditable?: AttributeBool;
      dir?: "ltr" | "rtl" | "auto";
      draggable?: AttributeBool | "auto";
      hidden?: AttributeBool;
      id?: string;
      inert?: AttributeBool;
      lang?: string;
      onabort?: string;
      onafterprint?: string;
      onbeforeprint?: string;
      onbeforeunload?: string;
      onblur?: string;
      oncanplay?: string;
      oncanplaythrough?: string;
      onchange?: string;
      onclick?: string;
      oncontextmenu?: string;
      oncopy?: string;
      oncuechange?: string;
      oncut?: string;
      ondblclick?: string;
      ondrag?: string;
      ondragend?: string;
      ondragenter?: string;
      ondragleave?: string;
      ondragover?: string;
      ondragstart?: string;
      ondrop?: string;
      ondurationchange?: string;
      onemptied?: string;
      onended?: string;
      onerror?: string;
      onfocus?: string;
      onhashchange?: string;
      oninput?: string;
      oninvalid?: string;
      onkeydown?: string;
      onkeypress?: string;
      onkeyup?: string;
      onload?: string;
      onloadeddata?: string;
      onloadedmetadata?: string;
      onloadstart?: string;
      onmousedown?: string;
      onmousemove?: string;
      onmouseout?: string;
      onmouseover?: string;
      onmouseup?: string;
      onmousewheel?: string;
      onoffline?: string;
      ononline?: string;
      onpagehide?: string;
      onpageshow?: string;
      onpaste?: string;
      onpause?: string;
      onplay?: string;
      onplaying?: string;
      onpopstate?: string;
      onprogress?: string;
      onratechange?: string;
      onreset?: string;
      onresize?: string;
      onscroll?: string;
      onsearch?: string;
      onseeked?: string;
      onseeking?: string;
      onselect?: string;
      onstalled?: string;
      onstorage?: string;
      onsubmit?: string;
      onsuspend?: string;
      ontimeupdate?: string;
      ontoggle?: string;
      onunload?: string;
      onvolumechange?: string;
      onwaiting?: string;
      onwheel?: string;
      role?: string;
      slot?: string;
      spellcheck?: AttributeBool;
      style?: string;
      tabindex?: string | number;
      title?: string;
      translate?: "yes" | "no";
    }
  }
}
