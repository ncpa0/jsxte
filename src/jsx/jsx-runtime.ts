import type { ComponentApi } from "../component-api/component-api";

type CreateElementProps = {
  [k: string]: any;
  children?: JSXTE.ElementChildren;
};

(() => {
  if(typeof Symbol.toHtmlTag === "symbol") return;

  Object.defineProperty(Symbol, "toHtmlTag", {
    value: Symbol("toHtmlTag"),
    enumerable: false,
    configurable: true,
  });
})();

declare global {
  interface SymbolConstructor {
    /**
     * Method under this property defines how an object should be converted to
     * HTML by the JSXTE renderer.
     *
     * @example
     *   class User {
     *     constructor(
     *       public firstname: string,
     *       public lastname: string,
     *       public email: string,
     *       public age: number,
     *     ) {}
     *
     *     [Symbol.toHtmlTag]() {
     *       return `${this.firstname} ${this.lastname}`;
     *     }
     *   }
     *
     *   const user = new User("John", "Doe", "joedoe@gmai.com", 26);
     *   renderToHtml(<div>{user}</div>); // => <div>John Doe</div>
     */
    readonly toHtmlTag: symbol;
  }
}

const mapChildren = (
  children: JSXTE.ElementChildren,
  accumulator: JSX.Element[],
) => {
  // eslint-disable-next-line @typescript-eslint/switch-exhaustiveness-check
  switch (typeof children) {
    case "string":
      accumulator.push({ type: "textNode", text: children });
      break;
    case "number":
      accumulator.push({ type: "textNode", text: children.toString() });
      break;
    case "object":
      if (Array.isArray(children)) {
        for (let i = 0; i < children.length; i++) {
          const child = children[i]!;
          mapChildren(child, accumulator);
        }
      } else if (children != null) {
        if (
          Symbol.toHtmlTag in children &&
          typeof children[Symbol.toHtmlTag] === "function"
        ) {
          const html = String((children[Symbol.toHtmlTag] as Function)());
          accumulator.push({ type: "textNode", text: html });
        } else {
          accumulator.push(children);
        }
      }
      break;
  }

  return accumulator;
};

export const createElement = (
  tag:
    | string
    | ((props: any, contextMap: ComponentApi) => JSX.Element)
    | ((props: any, contextMap: ComponentApi) => Promise<JSX.Element>),
  props?: CreateElementProps,
  ...children: Array<
    JSX.Element | string | number | Array<JSX.Element | string | number>
  >
): JSX.Element => {
  props ??= {};

  const finalChildren: JSX.Element[] = [];

  for (let i = 0; i < children.length; i++) {
    mapChildren(children[i]!, finalChildren);
  }

  if (props?.children) {
    mapChildren(props.children, finalChildren);
  }

  props.children = finalChildren;

  Object.freeze(finalChildren);
  Object.freeze(props);

  return {
    type: "tag",
    // @ts-expect-error
    tag,
    props: props as JSXTE.ElementProps,
  };
};

export const jsx = createElement;
export const jsxs = jsx;
export const _jsx = jsx;
export const _jsxs = jsx;

export const Fragment = "";
export const _Fragment = Fragment;
