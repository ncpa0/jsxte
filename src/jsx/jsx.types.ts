import type { HTMLProps } from "./base-html-tag-props";
import type { AnchorTagProps } from "./prop-types/a-jsx-props";
import type { AreaTagProps } from "./prop-types/area-jsx-props";
import type { AudioTagProps } from "./prop-types/audio-jsx-props";
import type { BaseTagProps } from "./prop-types/base-jsx-props";
import type { BdoTagProps } from "./prop-types/bdo-jsx-props";
import type { BlockquoteTagProps } from "./prop-types/blockquote-jsx-props";
import type { ButtonTagProps } from "./prop-types/button-jsx-props";
import type { CanvasTagProps } from "./prop-types/canvas-jsx-props";
import type { ColTagProps } from "./prop-types/col-jsx-props";
import type { ColgroupTagProps } from "./prop-types/colgroup-jsx-props";
import type { DataTagProps } from "./prop-types/data-jsx-props";
import type { DelTagProps } from "./prop-types/del-jsx-props";
import type { DetailsTagProps } from "./prop-types/details-jsx-props";
import type { DialogTagProps } from "./prop-types/dialog-jsx-props";
import type { EmbedTagProps } from "./prop-types/embed-jsx-props";
import type { FieldsetTagProps } from "./prop-types/fieldset-jsx-prop";
import type { FormTagProps } from "./prop-types/form-jsx-props";
import type { HtmlTagProps } from "./prop-types/html-jsx-props";
import type { IframeTagProps } from "./prop-types/iframe-jsx-props";
import type { ImgTagProps } from "./prop-types/img-jsx-props";
import type { InputTagProps } from "./prop-types/input-jsx-props";
import type { InsTagProps } from "./prop-types/ins-jsx-props";
import type { LabelTagProps } from "./prop-types/label-jsx-props";
import type { LiTagProps } from "./prop-types/li-jsx.props";
import type { LinkTagProps } from "./prop-types/link-jsx-props";
import type { MapTagProps } from "./prop-types/map-jsx-props";
import type { MetaTagProps } from "./prop-types/meta-jsx-props";
import type { MeterTagProps } from "./prop-types/meter-jsx-props";
import type { ObjectTagProps } from "./prop-types/object-jsx-props";
import type { OlTagProps } from "./prop-types/ol-jsx-props";
import type { OptgroupTagProps } from "./prop-types/optgroup-jsx-props";
import type { OptionTagProps } from "./prop-types/option-jsx-props";
import type { OutputTagProps } from "./prop-types/output-jsx-props";
import type { ParamTagProps } from "./prop-types/param-jsx-props";
import type { ProgressTagProps } from "./prop-types/progress-jsx-props";
import type { QTagProps } from "./prop-types/q-jsx-props";
import type { ScriptTagProps } from "./prop-types/script-jsx-props";
import type { SelectTagProps } from "./prop-types/select-jsx-props";
import type { SourceTagProps } from "./prop-types/source-jsx-props";
import type { StyleTagProps } from "./prop-types/style-jsx-props";
import type { TdTagProps } from "./prop-types/td-jsx-props";
import type { TextareaTagProps } from "./prop-types/textarea-jsx-props";
import type { ThTagProps } from "./prop-types/th-jsx-props";
import type { TimeTagProps } from "./prop-types/time-jsx-props";
import type { TrackTagProps } from "./prop-types/track-jsx-props";
import type { VideoTagProps } from "./prop-types/video-jsx-props";

declare global {
  namespace JSX {
    type AsyncElement = Promise<JSXTE.TagElement | JSXTE.TextNodeElement>;
    type SyncElement = JSXTE.TagElement | JSXTE.TextNodeElement;

    type Element = SyncElement | AsyncElement;

    interface IntrinsicElements {
      a: HTMLProps<AnchorTagProps>;
      abbr: HTMLProps;
      address: HTMLProps;
      area: HTMLProps<AreaTagProps>;
      article: HTMLProps;
      aside: HTMLProps;
      audio: HTMLProps<AudioTagProps>;
      b: HTMLProps;
      base: HTMLProps<BaseTagProps>;
      bdi: HTMLProps;
      bdo: HTMLProps<BdoTagProps>;
      big: HTMLProps;
      blockquote: HTMLProps<BlockquoteTagProps>;
      body: HTMLProps;
      br: HTMLProps;
      button: HTMLProps<ButtonTagProps>;
      canvas: HTMLProps<CanvasTagProps>;
      caption: HTMLProps;
      cite: HTMLProps;
      code: HTMLProps;
      col: HTMLProps<ColTagProps>;
      colgroup: HTMLProps<ColgroupTagProps>;
      data: HTMLProps<DataTagProps>;
      datalist: HTMLProps;
      dd: HTMLProps;
      del: HTMLProps<DelTagProps>;
      details: HTMLProps<DetailsTagProps>;
      dfn: HTMLProps;
      dialog: HTMLProps<DialogTagProps>;
      div: HTMLProps;
      dl: HTMLProps;
      dt: HTMLProps;
      em: HTMLProps;
      embed: HTMLProps<EmbedTagProps>;
      fieldset: HTMLProps<FieldsetTagProps>;
      figcaption: HTMLProps;
      figure: HTMLProps;
      footer: HTMLProps;
      form: HTMLProps<FormTagProps>;
      h1: HTMLProps;
      h2: HTMLProps;
      h3: HTMLProps;
      h4: HTMLProps;
      h5: HTMLProps;
      h6: HTMLProps;
      head: HTMLProps;
      header: HTMLProps;
      hgroup: HTMLProps;
      hr: HTMLProps;
      html: HTMLProps<HtmlTagProps>;
      i: HTMLProps;
      iframe: HTMLProps<IframeTagProps>;
      img: HTMLProps<ImgTagProps>;
      input: HTMLProps<InputTagProps>;
      ins: HTMLProps<InsTagProps>;
      kbd: HTMLProps;
      keygen: HTMLProps;
      label: HTMLProps<LabelTagProps>;
      legend: HTMLProps;
      li: HTMLProps<LiTagProps>;
      link: HTMLProps<LinkTagProps>;
      main: HTMLProps;
      map: HTMLProps<MapTagProps>;
      mark: HTMLProps;
      menu: HTMLProps;
      menuitem: HTMLProps;
      meta: HTMLProps<MetaTagProps>;
      meter: HTMLProps<MeterTagProps>;
      nav: HTMLProps;
      noindex: HTMLProps;
      noscript: HTMLProps;
      object: HTMLProps<ObjectTagProps>;
      ol: HTMLProps<OlTagProps>;
      optgroup: HTMLProps<OptgroupTagProps>;
      option: HTMLProps<OptionTagProps>;
      output: HTMLProps<OutputTagProps>;
      p: HTMLProps;
      param: HTMLProps<ParamTagProps>;
      picture: HTMLProps;
      pre: HTMLProps;
      progress: HTMLProps<ProgressTagProps>;
      q: HTMLProps<QTagProps>;
      rp: HTMLProps;
      rt: HTMLProps;
      ruby: HTMLProps;
      s: HTMLProps;
      samp: HTMLProps;
      slot: HTMLProps;
      script: HTMLProps<ScriptTagProps>;
      section: HTMLProps;
      select: HTMLProps<SelectTagProps>;
      small: HTMLProps;
      source: HTMLProps<SourceTagProps>;
      span: HTMLProps;
      strong: HTMLProps;
      style: HTMLProps<StyleTagProps>;
      sub: HTMLProps;
      summary: HTMLProps;
      sup: HTMLProps;
      table: HTMLProps;
      template: HTMLProps;
      tbody: HTMLProps;
      td: HTMLProps<TdTagProps>;
      textarea: HTMLProps<TextareaTagProps>;
      tfoot: HTMLProps;
      th: HTMLProps<ThTagProps>;
      thead: HTMLProps;
      time: HTMLProps<TimeTagProps>;
      title: HTMLProps;
      tr: HTMLProps;
      track: HTMLProps<TrackTagProps>;
      u: HTMLProps;
      ul: HTMLProps;
      var: HTMLProps;
      video: HTMLProps<VideoTagProps>;
      wbr: HTMLProps;
      webview: HTMLProps;

      // SVG
      svg: HTMLProps;

      // web components
    }
  }
}
