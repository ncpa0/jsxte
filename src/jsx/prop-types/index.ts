export type { AnchorTagProps } from "./a-jsx-props";
export type { AreaTagProps } from "./area-jsx-props";
export type { AudioTagProps } from "./audio-jsx-props";
export type { BaseTagProps } from "./base-jsx-props";
export type { BdoTagProps } from "./bdo-jsx-props";
export type { BlockquoteTagProps } from "./blockquote-jsx-props";
export type { ButtonTagProps } from "./button-jsx-props";
export type { CanvasTagProps } from "./canvas-jsx-props";
export type { ColTagProps } from "./col-jsx-props";
export type { ColgroupTagProps } from "./colgroup-jsx-props";
export type { DataTagProps } from "./data-jsx-props";
export type { DelTagProps } from "./del-jsx-props";
export type { DetailsTagProps } from "./details-jsx-props";
export type { DialogTagProps } from "./dialog-jsx-props";
export type { EmbedTagProps } from "./embed-jsx-props";
export type { FieldsetTagProps } from "./fieldset-jsx-prop";
export type { FormTagProps } from "./form-jsx-props";
export type { HtmlTagProps } from "./html-jsx-props";
export type { IframeTagProps } from "./iframe-jsx-props";
export type { ImgTagProps } from "./img-jsx-props";
export type { InputTagProps } from "./input-jsx-props";
export type { InsTagProps } from "./ins-jsx-props";
export type { LabelTagProps } from "./label-jsx-props";
export type { LiTagProps } from "./li-jsx.props";
export type { LinkTagProps } from "./link-jsx-props";
export type { MapTagProps } from "./map-jsx-props";
export type { MetaTagProps } from "./meta-jsx-props";
export type { MeterTagProps } from "./meter-jsx-props";
export type { ObjectTagProps } from "./object-jsx-props";
export type { OlTagProps } from "./ol-jsx-props";
export type { OptgroupTagProps } from "./optgroup-jsx-props";
export type { OptionTagProps } from "./option-jsx-props";
export type { OutputTagProps } from "./output-jsx-props";
export type { ParamTagProps } from "./param-jsx-props";
export type { ProgressTagProps } from "./progress-jsx-props";
export type { QTagProps } from "./q-jsx-props";
export type { ScriptTagProps } from "./script-jsx-props";
export type { SelectTagProps } from "./select-jsx-props";
export type { SourceTagProps } from "./source-jsx-props";
export type { StyleTagProps } from "./style-jsx-props";
export type { TdTagProps } from "./td-jsx-props";
export type { TextareaTagProps } from "./textarea-jsx-props";
export type { ThTagProps } from "./th-jsx-props";
export type { TimeTagProps } from "./time-jsx-props";
export type { TrackTagProps } from "./track-jsx-props";
export type { VideoTagProps } from "./video-jsx-props";
