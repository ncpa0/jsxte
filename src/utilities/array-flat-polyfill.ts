/* eslint-disable @typescript-eslint/prefer-for-of */

function flat<A extends any[], D extends number = 1>(
  this: A,
  depth?: D,
): FlatArray<A, D>[] {
  const copy: FlatArray<A, D>[] = [];

  if (depth === undefined) {
    // @ts-expect-error
    depth = 1;
  }

  for (let i = 0; i < this.length; i++) {
    const item = this[i];

    if (depth! > 0 && Array.isArray(item)) {
      const itemFlatCopy = flat.call(item, depth! - 1);
      for (let j = 0; j < itemFlatCopy.length; j++) {
        copy.push(itemFlatCopy[j]);
      }
    } else {
      copy.push(item);
    }
  }

  return copy;
}

// Add flat polyfill to Array prototype if not defined
if (!Array.prototype.flat) {
  Object.defineProperty(Array.prototype, "flat", {
    configurable: true,
    value: flat,
    writable: true,
  });
}
