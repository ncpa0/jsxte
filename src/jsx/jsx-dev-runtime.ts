import type { ComponentApi } from "../component-api/component-api";
import type { CreateElementProps } from "./jsx-runtime";
import { Fragment, jsx, jsxs } from "./jsx-runtime";

const jsxDEV = (
  tag:
    | string
    | ((props: any, contextMap: ComponentApi) => JSX.Element)
    | ((props: any, contextMap: ComponentApi) => Promise<JSX.Element>),
  props?: CreateElementProps,
) => {
  const { children, ...restProps } = props ?? {};
  if (Array.isArray(children)) {
    return jsx(tag, restProps, ...children);
  }
  return jsx(tag, restProps, children);
};

export { Fragment, jsx, jsxs, jsxDEV };
