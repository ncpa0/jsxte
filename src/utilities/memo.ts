import type { ContextMap } from "../context-map/context-map";
import { renderToHtml, renderToHtmlAsync } from "../html-parser/render-to-html";
import { jsx, Fragment } from "../jsx-runtime";
import { Cache } from "./cache";

const ReplaceMap = <P extends { context: ContextMap }>(
  props: JSXTE.PropsWithChildren<P>,
  context: ContextMap
) => {
  context.replace(props.context);
  // @ts-ignore
  return jsx(Fragment, {}, props.children);
};

/**
 * Creates an in-memory cache for the provided component and
 * returns a new component that will shallow compare props
 * provided to it and return a cached html string if the props
 * match those provided in previous renders.
 *
 * Props comparison is shallow, so it's best to avoid using
 * `memo` on components that have objects passed to it as props.
 *
 * Component children are not compared, so if only children
 * change, the component will use the cached html string and not
 * reflect those changes.
 */
export const memo = <P extends object & { children?: any }>(
  Component: (props: P, context: ContextMap) => JSX.Element,
  options?: {
    /** Time in milliseconds. Default: 15 minutes. */
    maxCacheAge?: number;
    /**
     * Whether to use `renderToHtmlAsync` or `renderToHtml` to
     * render this component. Default: `false`.
     */
    renderAsynchronously?: boolean;
  }
) => {
  const { maxCacheAge, renderAsynchronously = false } = options ?? {};

  const cache = new Cache<JSXTE.TextNodeElement>(maxCacheAge);

  if (renderAsynchronously) {
    const MemoComponentAsync = async (
      props: P,
      context: ContextMap
    ): Promise<JSXTE.TextNodeElement> => {
      const { children, ...propsNoChildren } = props;

      const cachedResult = cache.get(propsNoChildren);
      if (cachedResult) return cachedResult;

      const result = await renderToHtmlAsync(
        jsx(
          ReplaceMap,
          { context },
          jsx(Component, { ...propsNoChildren }, children)
        )
      );
      const textNode: JSXTE.TextNodeElement = {
        text: result,
        type: "textNode",
      };
      cache.set(propsNoChildren, textNode);

      return textNode;
    };

    return MemoComponentAsync;
  }

  const MemoComponent = (
    props: P,
    context: ContextMap
  ): JSXTE.TextNodeElement => {
    const { children, ...propsNoChildren } = props;

    const cachedResult = cache.get(propsNoChildren);
    if (cachedResult) return cachedResult;

    const result = renderToHtml(
      jsx(
        ReplaceMap,
        { context },
        jsx(Component, { ...propsNoChildren }, children)
      )
    );
    const textNode: JSXTE.TextNodeElement = {
      text: result,
      type: "textNode",
    };
    cache.set(propsNoChildren, textNode);

    return textNode;
  };

  return MemoComponent;
};
