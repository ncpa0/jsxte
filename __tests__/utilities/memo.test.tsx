import { describe, expect, it } from "vitest";
import {
  type ComponentApi,
  defineContext,
} from "../../src/component-api/component-api";
import {
  renderToHtml,
  renderToHtmlAsync,
} from "../../src/html-renderer/render-to-html";
import { jsx } from "../../src/jsx/jsx-runtime";
import { memo } from "../../src/utilities/memo";

const sleep = (t: number) =>
  new Promise<void>((resolve) => setTimeout(() => resolve(), t));

describe("memo", () => {
  it("should correctly memoize the component", () => {
    let variable = "Hello World";

    const MemoizedComponent = memo(() => {
      return <h1>{variable}</h1>;
    });

    const Root = () => {
      return (
        <div id="root">
          <MemoizedComponent />
        </div>
      );
    };

    const render1 = renderToHtml(<Root />);

    variable = "Bye cruel world";

    const render2 = renderToHtml(<Root />);

    expect(render1).toEqual(render2);
    expect(render1).toMatchSnapshot();
  });

  it("should \"forget\" memoized results that are past expire time", async () => {
    let variable = "Hello World";

    const MemoizedComponent = memo(
      () => {
        return <h1>{variable}</h1>;
      },
      { maxCacheAge: 100 },
    );

    const Root = () => {
      return (
        <div id="root">
          <MemoizedComponent />
        </div>
      );
    };

    const render1 = renderToHtml(<Root />);

    variable = "Bye cruel world";

    const render2 = renderToHtml(<Root />);

    expect(render1).toEqual(render2);

    await sleep(150);

    const render3 = renderToHtml(<Root />);

    expect(render1).not.toEqual(render3);
    expect(render3).toMatchSnapshot();
  });

  it("memoized components should have access to the context of it's parent", () => {
    const Context = defineContext<{ foo: string }>();

    const MemoizedComponent = memo((_: {}, { ctx }) => {
      return <h1>{ctx.getOrFail(Context).foo}</h1>;
    });

    const Root = (_: {}, { ctx }: ComponentApi) => {
      ctx.set(Context, { foo: "Hello" });

      return (
        <div id="root">
          <MemoizedComponent />
        </div>
      );
    };

    const render1 = renderToHtml(<Root />);

    expect(render1).toMatchSnapshot();
  });

  it("children of the memoized components should have access to the top level context", () => {
    const Context = defineContext<{ foo: string }>();

    const MemoChild = (_: {}, { ctx }: ComponentApi) => {
      return <h1>{ctx.getOrFail(Context).foo}</h1>;
    };

    const MemoizedComponent = memo((_: {}, context) => {
      return (
        <div>
          <MemoChild />
        </div>
      );
    });

    const MemoParent = (_: {}, context: ComponentApi) => {
      return (
        <div>
          <MemoizedComponent />
        </div>
      );
    };

    const Root = (_: {}, { ctx }: ComponentApi) => {
      ctx.set(Context, { foo: "Hello" });

      return (
        <div id="root">
          <MemoParent />
        </div>
      );
    };

    const render1 = renderToHtml(<Root />);

    expect(render1).toMatchSnapshot();
  });

  it("asynchronous memoized components should works as expected", async () => {
    let variable = "Hello World";

    const MemoizedComponent = memo(
      async () => {
        await sleep(100);

        return <h1>{variable}</h1>;
      },
      { renderAsynchronously: true },
    );

    const Root = () => {
      return (
        <div id="root">
          <MemoizedComponent />
        </div>
      );
    };

    const render1 = await renderToHtmlAsync(<Root />);

    variable = "Bye cruel world";

    const render2 = await renderToHtmlAsync(<Root />);

    expect(render1).toEqual(render2);
    expect(render1).toMatchSnapshot();
  });

  it("memoized component should still re-render when props change", () => {
    const MemoizedComponent = memo((props: { foo: string }) => {
      return <h1>{props.foo}</h1>;
    });

    const Root = (props: { foo: string }) => {
      return (
        <div id="root">
          <MemoizedComponent foo={props.foo} />
        </div>
      );
    };

    const render1 = renderToHtml(<Root foo="a" />);

    const render2 = renderToHtml(<Root foo="b" />);

    expect(render1).not.toEqual(render2);
  });
});
