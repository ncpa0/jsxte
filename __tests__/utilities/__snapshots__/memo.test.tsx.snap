// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`memo > asynchronous memoized components should works as expected 1`] = `
"<div id="root">
  <h1>
    Hello World
  </h1>
</div>
"
`;

exports[`memo > children of the memoized components should have access to the top level context 1`] = `
"<div id="root">
  <div>
    <div>
      <h1>
        Hello
      </h1>
    </div>
  </div>
</div>
"
`;

exports[`memo > memoized components should have access to the context of it's parent 1`] = `
"<div id="root">
  <h1>
    Hello
  </h1>
</div>
"
`;

exports[`memo > should "forget" memoized results that are past expire time 1`] = `
"<div id="root">
  <h1>
    Bye cruel world
  </h1>
</div>
"
`;

exports[`memo > should correctly memoize the component 1`] = `
"<div id="root">
  <h1>
    Hello World
  </h1>
</div>
"
`;
